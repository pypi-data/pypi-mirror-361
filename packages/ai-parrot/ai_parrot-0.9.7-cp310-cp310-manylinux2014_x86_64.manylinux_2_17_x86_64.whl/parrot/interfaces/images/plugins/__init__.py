"""
Image processing and generation interfaces for Parrot.
"""
from .yolo import YOLOPlugin
from .vision import VisionTransformerPlugin
from .hash import ImageHashPlugin
from .abstract import ImagePlugin
from .exif import EXIFPlugin
from .zerodetect import ZeroShotDetectionPlugin


PLUGINS = {
    "exif": EXIFPlugin,
    "hash": ImageHashPlugin,
    "yolo": YOLOPlugin,
    "vectorization": VisionTransformerPlugin,
    "zeroshot": ZeroShotDetectionPlugin,
}
