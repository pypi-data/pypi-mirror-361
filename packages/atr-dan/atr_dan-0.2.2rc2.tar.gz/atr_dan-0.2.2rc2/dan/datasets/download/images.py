# Copyright Teklia (contact@teklia.com) & Denis Coquenet
# This code is licensed under CeCILL-C

# -*- coding: utf-8 -*-

import functools
import json
import logging
import pickle
from collections import defaultdict
from concurrent.futures import Future, ThreadPoolExecutor
from itertools import chain
from pathlib import Path
from typing import Dict, List, Tuple

import cv2
import numpy as np
from PIL import Image
from tqdm import tqdm

from dan import TRAIN_NAME
from dan.datasets.download.exceptions import ImageDownloadError
from dan.datasets.download.utils import (
    download_image,
    get_bbox,
)
from line_image_extractor.extractor import extract
from line_image_extractor.image_utils import (
    BoundingBox,
    Extraction,
    polygon_to_bbox,
)

IMAGES_DIR = "images"  # Subpath to the images directory.

IIIF_URL = "{image_url}/{bbox}/{size}/0/default.jpg"
# IIIF 2.0 uses `full`
IIIF_FULL_SIZE = "full"

logger = logging.getLogger(__name__)


def output_attr_required(func):
    """
    Always check that the output attribute is not null.
    """

    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        assert self.output is not None, "Define an output folder to download images."
        return func(self, *args, **kwargs)

    return wrapper


class ImageDownloader:
    """
    Download images from extracted data
    """

    def __init__(
        self,
        output: Path | None = None,
        max_width: int | None = None,
        max_height: int | None = None,
        image_extension: str = "",
        unknown_token: str = "⁇",
    ) -> None:
        self.output: Path | None = output

        self.max_width: int | None = max_width
        self.max_height: int | None = max_height
        self.image_extension: str = image_extension
        self.data: dict[str, dict] = defaultdict(dict)
        self.unknown_token: str = unknown_token

    @output_attr_required
    def load_split_data(self) -> None:
        """
        Load the dataset stored in `split.json` and initializes the charset.
        """
        # Load split file
        split_file = self.output / "split.json"
        self.split: Dict = (
            json.loads(split_file.read_text())
            if split_file and split_file.is_file()
            else {}
        )

        # Create directories
        for split_name in self.split:
            (self.output / IMAGES_DIR / split_name).mkdir(parents=True, exist_ok=True)

        self.charset = set(
            chain.from_iterable(
                split_data["text"] for split_data in self.split[TRAIN_NAME].values()
            )
        )

        # Add unknown token to charset
        self.charset.add(self.unknown_token)

    def check_extraction(self, values: dict) -> str | None:
        # Check dataset_id parameter
        if values.get("dataset_id") is None:
            return "Dataset ID not found"

        # Check image parameters
        if not (image := values.get("image")):
            return "Image information not found"

        # Only support iiif_url with polygon for now
        if not image.get("iiif_url"):
            return "Image IIIF URL not found"
        if not image.get("polygon"):
            return "Image polygon not found"

        # Check text parameter
        if values.get("text") is None:
            return "Text not found"

        if self.unknown_token in values["text"]:
            return "Unknown token found in the transcription text"

    def get_iiif_size_arg(self, width: int, height: int) -> str:
        if (self.max_width is None or width <= self.max_width) and (
            self.max_height is None or height <= self.max_height
        ):
            return IIIF_FULL_SIZE

        bigger_width = self.max_width and width >= self.max_width
        bigger_height = self.max_height and height >= self.max_height

        if bigger_width and bigger_height:
            # Resize to the biggest dim to keep aspect ratio
            # Only resize width is bigger than max size
            # This ratio tells which dim needs the biggest shrinking
            ratio = width * self.max_height / (height * self.max_width)
            return f"{self.max_width}," if ratio > 1 else f",{self.max_height}"
        elif bigger_width:
            return f"{self.max_width},"
        # Only resize height is bigger than max size
        elif bigger_height:
            return f",{self.max_height}"

    def build_iiif_url(
        self, polygon: List[List[int]], image_url: str
    ) -> Tuple[BoundingBox, str]:
        bbox = polygon_to_bbox(polygon)
        size = self.get_iiif_size_arg(width=bbox.width, height=bbox.height)
        # Rotations are done using the lib
        return IIIF_URL.format(image_url=image_url, bbox=get_bbox(polygon), size=size)

    @output_attr_required
    def build_tasks(self) -> List[Dict[str, str]]:
        tasks = []
        for split, items in self.split.items():
            # Create directories
            destination = self.output / IMAGES_DIR / split
            destination.mkdir(parents=True, exist_ok=True)

            for element_id, values in items.items():
                filename = Path(element_id).with_suffix(self.image_extension)

                error = self.check_extraction(values)
                if error:
                    logger.warning(f"{destination / filename}: {error}")
                    continue

                image_path = destination / values["dataset_id"] / filename
                image_path.parent.mkdir(parents=True, exist_ok=True)

                # Replace unknown characters by the unknown token
                if split != TRAIN_NAME:
                    unknown_charset = set(values["text"]) - self.charset
                    values["text"] = values["text"].translate(
                        {
                            ord(unknown_char): self.unknown_token
                            for unknown_char in unknown_charset
                        }
                    )

                # Store a relative path to the label file in case we need to move the data elsewhere
                self.data[split][str(image_path.relative_to(self.output))] = values[
                    "text"
                ]

                # Create task for multithreading pool if image does not exist yet
                if image_path.exists():
                    continue

                polygon = values["image"]["polygon"]
                iiif_url = values["image"]["iiif_url"]
                tasks.append(
                    {
                        "split": split,
                        "polygon": polygon,
                        "image_url": self.build_iiif_url(polygon, iiif_url),
                        "destination": image_path,
                    }
                )
        return tasks

    def get_image(
        self,
        split: str,
        polygon: List[List[int]],
        image_url: str,
        destination: Path,
    ) -> None:
        """Save the element's image to the given path and applies any image operations needed.

        :param split: Dataset split this image belongs to.
        :param polygon: Polygon of the processed element.
        :param image_url: Base IIIF URL of the image.
        :param destination: Where the image should be saved.
        """
        bbox = polygon_to_bbox(polygon)
        try:
            img: Image.Image = download_image(image_url)

            # The polygon's coordinate are in the referential of the full image
            # We need to remove the offset of the bounding rectangle
            polygon = [(x - bbox.x, y - bbox.y) for x, y in polygon]

            # Normalize bbox
            bbox = BoundingBox(x=0, y=0, width=bbox.width, height=bbox.height)

            image = extract(
                img=cv2.cvtColor(np.asarray(img), cv2.COLOR_RGB2BGR),
                polygon=np.asarray(polygon).clip(0),
                bbox=bbox,
                extraction_mode=Extraction.boundingRect,
                max_deskew_angle=45,
            )

            cv2.imwrite(str(destination), image)

        except Exception as e:
            raise ImageDownloadError(
                split=split, path=destination, url=image_url, exc=e
            )

    def download_images(self, tasks: List[Dict[str, str]]) -> None:
        """
        Execute each image download task in parallel

        :param tasks: List of tasks to execute.
        """
        failed_downloads = []
        with (
            tqdm(desc="Downloading images", total=len(tasks)) as pbar,
            ThreadPoolExecutor() as executor,
        ):

            def process_future(future: Future):
                """
                Callback function called at the end of the thread
                """
                # Update the progress bar count
                pbar.update(1)

                exc = future.exception()
                if exc is None:
                    # No error
                    return
                # If failed, tag for removal
                assert isinstance(exc, ImageDownloadError)
                # Remove transcription from labels dict
                del self.data[exc.split][exc.path]
                # Save tried URL
                failed_downloads.append((exc.url, exc.message))

            # Submit all tasks
            for task in tasks:
                executor.submit(self.get_image, **task).add_done_callback(
                    process_future
                )

        if failed_downloads:
            logger.error(f"Failed to download {len(failed_downloads)} image(s).")
            print(*list(map(": ".join, failed_downloads)), sep="\n")

    @output_attr_required
    def export(self) -> None:
        """
        Writes a `labels.json` file containing a mapping of the images that have been correctly uploaded (identified by its path)
        to the ground-truth transcription (with NER tokens if needed).
        """
        (self.output / "labels.json").write_text(
            json.dumps(
                self.data,
                sort_keys=True,
                indent=4,
            )
        )

        (self.output / "charset.pkl").write_bytes(
            pickle.dumps(sorted(list(self.charset)))
        )

    @output_attr_required
    def run(self) -> None:
        """
        Download the missing images from a `split.json` file and build a `labels.json` file containing
        a mapping of the images that have been correctly uploaded (identified by its path)
        to the ground-truth transcription (with NER tokens if needed).
        """
        self.load_split_data()
        tasks: list[dict[str, str]] = self.build_tasks()
        self.download_images(tasks)
        self.export()


def run(
    output: Path,
    max_width: int | None,
    max_height: int | None,
    image_format: str,
    unknown_token: str,
):
    """
    Download the missing images from a `split.json` file and build a `labels.json` file containing
    a mapping of the images that have been correctly uploaded (identified by its path)
    to the ground-truth transcription (with NER tokens if needed).

    :param output: Path where the `split.json` file is stored and where the data will be generated
    :param max_width: Images larger than this width will be resized to this width
    :param max_height: Images larger than this height will be resized to this height
    :param image_format: Images will be saved under this format
    :param unknown_token: The token used to replace unknown characters.
    """
    ImageDownloader(
        output=output,
        max_width=max_width,
        max_height=max_height,
        image_extension=image_format,
        unknown_token=unknown_token,
    ).run()
