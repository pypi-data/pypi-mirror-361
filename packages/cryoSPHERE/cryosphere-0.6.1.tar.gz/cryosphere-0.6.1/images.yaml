apix: 1.0 #Size of a pixel in Ã…
Npix_downsize: 190 #Number of pixel along one dimension in the downsized image. Note that Npix_downsize == Npix no downsampling is applied.
Npix: 190 #Number of pixel along one dimension of the images
sigma_gmm: 2 #Sigma used in the Gaussian Mixture representation of the volume, see paper. We advise not to change it.
