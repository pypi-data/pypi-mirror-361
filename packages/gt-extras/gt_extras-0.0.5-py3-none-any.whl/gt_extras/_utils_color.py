from __future__ import annotations

from great_tables._data_color.palettes import GradientPalette
from great_tables._data_color.constants import DEFAULT_PALETTE, ALL_PALETTES
from great_tables._data_color.base import (
    _html_color,
    _rescale_factor,
    _get_domain_factor,
)

from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None, data: list, data_table: TblData
) -> list[str]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table, vals=data, domain=domain_factor, palette=palette
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    return color_vals
