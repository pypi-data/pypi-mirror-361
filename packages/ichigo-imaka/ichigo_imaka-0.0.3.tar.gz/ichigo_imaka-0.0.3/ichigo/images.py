"""Functions for importing and manipulating images.
"""

import os
import glob
import numpy as np

from astropy.io import fits
from numpy.typing import NDArray
from scipy.interpolate import RectBivariateSpline
from PIL import Image


def get_image(fn: str) -> NDArray:
    """Returns an image file as a numpy array.

    Parameters
    ----------
    fn: str
        Path to the image.

    Returns
    -------
    out: nd_array of size (n, m)
        A 2D array containing the PSF.
    """
    if fn.lower().endswith('.fits'):
        img = fits.getdata(fn, ext=0)
    else:
        try:
            img = Image.open(fn).convert('L')  # convert to grayscale
        except:
            raise ValueError("File type cannot be read as image.")
    return np.array(img)


def get_images_in_path(path: str, recursive: bool = False, extensions: list[str] = ["fits"]) -> NDArray:
    """Returns an array containing all of the images in the specified directory.
    All images must have the same shape.

    Parameters
    ----------
    path: str
        Directory containing images to combine.
    recursive: bool, optional
        If True, the directory is searched recursively.
    extensions: list of str, optional
        List of file extensions to include (upper or lowercase). Default is ["fits"].

    Returns
    -------
    out: nd_array of size (k, n, m)
        Array containing k images of size (n, m),
    """
    fnames = []
    for ext in extensions:
        # Search for both lower and upper case extensions
        queries = [os.path.join(path, '*.' + e) for e in (ext.lower(), ext.upper())]
        results = [glob.glob(query, recursive=recursive) for query in queries]
        fnames += results[0] + results[1]

    images = []
    for fname in fnames:
        img = get_image(fname)
        images.append(img)
    return np.array(images)


def stack_images(path: str, ext: str, path_bg: str | None = None, fn_out: str | None = None) -> NDArray:
    """Returns a stacked (summed) image from the images in the specified directory.
    All images must have the same dimensions.

    Parameters
    ----------
    path: str
        Directory containing images to combine.
    ext: str
        File extension of the images to stack.
    path_bg: str or None, optional
        Path to background images. If not None, this function creates a background
        frame by averaging all of the images in path_bg. This frame is subtracted
        from each image before stacking.
    fn_out: str or None, optional
        File name to save the stacked image to, including the file extension. If
        None, the file is not saved.

    Returns
    -------
    out: nd_array
        The stacked image.
    """
    imgs = get_images_in_path(path, extensions=[ext])

    if path_bg is not None:
        imgs_bg = get_images_in_path(path_bg)
        bg_averaged = np.average(imgs_bg, axis=0)
        imgs = imgs - bg_averaged[np.newaxis, :, :]

    img_stack = np.sum(imgs, axis=0)

    if fn_out:
        ext = os.path.splitext(fn_out)[1]
        ext = ext.lower()
        if ext == ".fits":
            hdu = fits.PrimaryHDU(img_stack)
            hdu.header['comment'] = "Averaged over the files in " + path
            hdu.writeto(fn_out, overwrite=True)
        else:
            img_to_save = Image.fromarray(img_stack)
            img_to_save.save(fn_out)
    
    return img_stack
    

def upsample_image(img: NDArray, s: int) -> NDArray:
    """Returns an image upsampled via 2D b-spline interpolation.

    Parameters
    ----------
    img: nd_array of size (n, m)
        Image to upsample.
    s: int
        Factor to upsample the image by.

    Returns
    -------
    out: nd_array of size (n*s, m*s)
        Upsampled image.
    """
    # x and y are flipped in RectBivariateSpline... They seem to be i and j...
    xi = np.arange(img.shape[0])
    yi = np.arange(img.shape[1])
    # interp2d is depreciated as of scipy 1.10.0. RegularGridInterpolator is recommended
    # for grid data but it is *excruciatingly* slow so I am using
    # RectBivariateSpline instead
    interp = RectBivariateSpline(xi, yi, img)
    xnew = np.linspace(xi[0], xi[-1], xi[-1]*s)
    ynew = np.linspace(yi[0], yi[-1], yi[-1]*s)
    return interp(xnew, ynew)