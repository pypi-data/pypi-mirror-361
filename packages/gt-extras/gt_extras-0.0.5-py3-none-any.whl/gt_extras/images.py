from __future__ import annotations

from great_tables import html
from great_tables._text import Html

__all__ = ["img_header"]


def img_header(
    label: str,
    img_url: str,
    height: float = 60,
    font_size: int = 12,
    border_color: str = "black",
    text_color: str = "black",
) -> Html:
    """
    Create an HTML header with an image and a label, apt for a column label.

    Parameters
    ----------
    label
        The text label to display below the image.

    img_url
        The URL of the image to display. This can be a filepath or an image on the web.

    height
        The height of the image in pixels.

    font_size
        The font size of the label text.

    border_color
        The color of the border below the image.

    text_color
        The color of the label text.

    Returns
    -------
    html
        A Great Tables `html` element for the header.

    Examples
    -------
    ```{python}
    import pandas as pd
    from great_tables import GT, md
    import gt_extras as gte

    df = pd.DataFrame(
        {
            "Category": ["Points", "Rebounds", "Assists", "Blocks", "Steals"],
            "Hart": [1051, 737, 453, 27, 119],
            "Brunson": [1690, 187, 475, 8, 60],
            "Bridges": [1444, 259, 306, 43, 75],
        }
    )

    hart_header = gte.img_header(
        label="Josh Hart",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3062679.png",
    )

    brunson_header = gte.img_header(
        label="Jalen Brunson",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3934672.png",
    )

    bridges_header = gte.img_header(
        label="Mikal Bridges",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3147657.png",
    )

    (
        GT(df, rowname_col="Category")
        .tab_source_note(md("Images and data courtesy of [ESPN](https://www.espn.com)"))
        .cols_label(
            {
                "Hart": hart_header,
                "Brunson": brunson_header,
                "Bridges": bridges_header,
            }
        )
    )
    ```
    """

    img_html = f"""
    <img src="{img_url}" style="
        height:{height}px;
        border-bottom:2px solid {border_color};"
    />
    """.strip()

    label_html = f"""
    <div style="
        font-size:{font_size}px;
        color:{text_color};
        text-align:center;
        width:100%;
    ">
        {label}
    </div>
    """.strip()

    full_element = f"""
    <div style="text-align:center;">
        {img_html}
        {label_html}
    </div>
    """.strip()

    return html(full_element)
