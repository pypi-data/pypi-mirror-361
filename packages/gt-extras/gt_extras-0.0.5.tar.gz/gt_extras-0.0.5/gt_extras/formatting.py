from __future__ import annotations

import pandas as pd
import polars as pl
from great_tables import GT
from great_tables._gt_data import Boxhead, ColInfo
from great_tables._tbl_data import SelectExpr, copy_frame, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["fmt_pct_extra", "gt_duplicate_column"]


def fmt_pct_extra(
    gt: GT,
    columns: SelectExpr,
    scale: float = 100,
    threshold: float = 1.0,
    color: str = "grey",
    decimals: int = 1,
) -> GT:
    """
    Convert to percent and show less than `1%` as `<1%` in grey.

    The `fmt_pct_extra()` function takes an existing `GT` object and formats a column of numeric
    values as percentages. Values below the specified threshold are displayed as `"<threshold%"`
    instead of their actual percentage value, and in a unique color.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns containing numeric values to format as percentages.

    scale
        Multiplication factor to convert values to percentages.
        Use `100` if values are decimals `(0.05 -> 5%)` (default),
        use `1` if values are already percentages `(5 -> 5%)`.

    threshold
        The percentage threshold below which values are displayed as `"<threshold%"` instead of
        their actual value. Note this refers to the scaled value, not the original.

    color
        The color to use for values below the threshold.

    decimals
        Number of decimal places to display for percentages.

    Returns
    -------
    GT
        A `GT` object with formatted percentage column.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import towny
    import gt_extras as gte

    towny_mini = towny[
        [
            "name",
            "pop_change_1996_2001_pct",
            "pop_change_2001_2006_pct",
            "pop_change_2006_2011_pct",
        ]
    ].tail(10)

    gt = (
        GT(towny_mini)
        .tab_spanner(label="Population Change", columns=[1, 2, 3])
        .cols_label(
            pop_change_1996_2001_pct="'96-'01",
            pop_change_2001_2006_pct="'01-'06",
            pop_change_2006_2011_pct="'06-'11",
        )
    )

    gt.pipe(
        gte.fmt_pct_extra,
        columns=[1, 2, 3],
        threshold=5,
        color="green",
    )
    ```
    """
    # TODO: consider how to handle negative values

    def _fmt_pct_single_val(value: float):
        if is_na(gt._tbl_data, value):
            return ""

        # Convert to percentage
        pct_value = value * scale

        if abs(pct_value) < threshold:
            return f"<span style='color:{color};'><{threshold:g}%</span>"
        else:
            return f"{pct_value:.{decimals}f}%"

    res = gt
    res = res.fmt(_fmt_pct_single_val, columns=columns)

    return res


def gt_duplicate_column(
    gt: GT,
    column: SelectExpr,
    after: str | None = None,
    append_text: str | None = "_dupe",
    dupe_name: str | None = None,
) -> GT:
    """
    Duplicate a column in a `GT` object.

    The `gt_duplicate_column()` function takes an existing `GT` object and creates a duplicate
    (without styling) of the specified column. The duplicated column can be renamed using either
    `dupe_name` or by appending text to the original column name, and positioned at a
    specific location in the table.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    column
        The column to duplicate. Can be a column name or index.

    after
        The column after which to place the duplicated column. If `None`, the duplicated
        column will be moved to the end of the table.

    append_text
        Text to append to the original column name for the duplicate. Only used if
        `dupe_name` is not provided. Defaults to `"_dupe"`.

    dupe_name
        The name for the duplicated column. If provided, this overrides `append_text`.

    Returns
    -------
    GT
        A `GT` object with the duplicated column added.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["mfr", "model", "year", "hp"]].head(5)
    gt = GT(gtcars_mini)

    # Duplicate with custom name and position
    gt.pipe(
        gte.gt_duplicate_column,
        column="hp",
        after="year",
    )
    ```
    """
    original_name, _ = _validate_and_get_single_column(gt, column)

    # If dupe_name is given, it overrides append_text
    append_text = append_text or "_dupe"
    if dupe_name is not None:
        new_col_name = dupe_name
    else:
        new_col_name = original_name + append_text

    if new_col_name == original_name:
        raise ValueError(
            f"The new column name '{new_col_name}' cannot be the same as the original column name '{original_name}'."
        )

    res = gt
    new_data_table = copy_frame(res._tbl_data)
    new_body = res._body.copy()

    # get the boxhead info
    original_col_info = None
    for col_info in res._boxhead:
        if col_info.var == original_name:
            original_col_info = col_info
            break

    # make the new boxhead entry
    new_col_info = ColInfo(
        var=new_col_name,
        type=original_col_info.type,
        column_label=new_col_name,
        column_align=original_col_info.column_align,
        column_width=original_col_info.column_width,
    )

    # A little clunky, but I dont have any other solutions
    if isinstance(new_data_table, pd.DataFrame):
        new_data_table[new_col_name] = new_data_table[original_name]
        new_body.body[new_col_name] = new_body.body[original_name]

    elif isinstance(new_data_table, pl.DataFrame):
        new_data_table = new_data_table.with_columns(
            new_data_table[original_name].alias(new_col_name)
        )
        new_body.body = new_body.body.with_columns(
            new_body.body[original_name].alias(new_col_name)
        )

    else:
        raise TypeError(
            """Unsupported type.
            This function will only work if the underlying data is a Polars or Pandas dataframe."""
        )

    new_boxhead_list = list(res._boxhead._d) + [new_col_info]
    new_boxhead = Boxhead(new_boxhead_list)

    res = res._replace(_tbl_data=new_data_table, _boxhead=new_boxhead, _body=new_body)

    if after is None:
        res = res.cols_move_to_end(new_col_name)
    else:
        res = res.cols_move(new_col_name, after=after)

    return res
