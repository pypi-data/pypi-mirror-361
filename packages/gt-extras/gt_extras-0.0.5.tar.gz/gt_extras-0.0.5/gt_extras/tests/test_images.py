from great_tables._text import Html
from gt_extras.images import img_header


def test_img_header_snapshot(snapshot):
    result = img_header(label="Test Label", img_url="https://example.com/image.png")
    assert snapshot == result


def test_img_header_basic():
    result = img_header(label="Test Label", img_url="https://example.com/image.png")

    assert isinstance(result, Html)
    assert "Test Label" in result.text
    assert "https://example.com/image.png" in result.text
    assert "height:60px;" in result.text
    assert "border-bottom:2px solid black;" in result.text
    assert "color:black;" in result.text


def test_img_header_custom_height_and_colors():
    result = img_header(
        label="Custom Label",
        img_url="https://example.com/custom.png",
        height=100,
        border_color="blue",
        text_color="red",
    )

    assert isinstance(result, Html)
    assert "Custom Label" in result.text
    assert "https://example.com/custom.png" in result.text
    assert "height:100px;" in result.text
    assert "border-bottom:2px solid blue;" in result.text
    assert "color:red;" in result.text


def test_img_header_custom_font_size():
    result = img_header(
        label="Font Size Test", img_url="https://example.com/font.png", font_size=20
    )

    assert isinstance(result, Html)
    assert "Font Size Test" in result.text
    assert "font-size:20px;" in result.text


def test_img_header_empty_label():
    result = img_header(label="", img_url="https://example.com/empty_label.png")

    assert isinstance(result, Html)
    assert "https://example.com/empty_label.png" in result.text
    assert "<div" in result.text
    assert "font-size:12px;" in result.text


def test_img_header_empty_url():
    result = img_header(label="Invalid URL Test", img_url="")

    assert isinstance(result, Html)
    assert "Invalid URL Test" in result.text
    assert 'src=""' in result.text


def test_img_header_no_border():
    result = img_header(
        label="No Border Test",
        img_url="https://example.com/no_border.png",
        border_color="transparent",
    )

    assert isinstance(result, Html)
    assert "No Border Test" in result.text
    assert "border-bottom:2px solid transparent;" in result.text
