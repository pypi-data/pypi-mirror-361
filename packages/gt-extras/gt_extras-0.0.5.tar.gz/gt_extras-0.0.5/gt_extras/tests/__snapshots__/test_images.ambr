# serializer version: 1
# name: test_img_header_snapshot
  Html(text='<div style="text-align:center;">\n        <img src="https://example.com/image.png" style="\n        height:60px;\n        border-bottom:2px solid black;"\n    />\n        <div style="\n        font-size:12px;\n        color:black;\n        text-align:center;\n        width:100%;\n    ">\n        Test Label\n    </div>\n    </div>')
# ---
